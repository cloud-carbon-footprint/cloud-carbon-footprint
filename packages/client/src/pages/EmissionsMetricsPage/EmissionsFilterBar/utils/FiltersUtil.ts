/*
 * Â© 2021 Thoughtworks, Inc.
 */

import { DropdownFilterOptions, DropdownOption, FilterOptions } from 'Types'
import createOptionChooser from './options'

export type DropdownSelections = {
  [key in DropdownFilterOptions]: DropdownOption[]
}

export function handleDropdownSelections(
  filterType: DropdownFilterOptions,
  selections: DropdownOption[],
  oldSelections: DropdownSelections,
  filterOptions: FilterOptions,
): DropdownSelections {
  return createOptionChooser(
    filterType,
    selections,
    oldSelections,
    filterOptions,
  ).choose()
}

export function numSelectedLabel(
  length: number,
  totalLength: number,
  type = 'Services',
): string {
  const lengthWithoutAllOption = totalLength - 1
  if (length === totalLength) {
    return `${type}: ${lengthWithoutAllOption} of ${lengthWithoutAllOption}`
  } else {
    return `${type}: ${length} of ${lengthWithoutAllOption}`
  }
}
