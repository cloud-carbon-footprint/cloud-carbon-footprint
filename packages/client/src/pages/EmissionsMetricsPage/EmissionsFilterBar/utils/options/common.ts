/*
 * Â© 2021 Thoughtworks, Inc.
 */

import { DropdownOption } from 'Types'

export function isDropdownOptionInDropdownOptions(
  comparingDropdownOptions: DropdownOption[],
  dropdownOption: DropdownOption,
): boolean {
  let isWithinComparingDropdownOption = false
  comparingDropdownOptions.forEach((comparingDropdownOption) => {
    if (comparingDropdownOption.key === dropdownOption.key) {
      isWithinComparingDropdownOption = true
    }
  })
  return isWithinComparingDropdownOption
}
