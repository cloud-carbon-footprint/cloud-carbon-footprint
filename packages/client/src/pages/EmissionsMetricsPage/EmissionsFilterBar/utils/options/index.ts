/*
 * Â© 2021 Thoughtworks, Inc.
 */

import { DropdownOption, FilterOptions, DropdownFilterOptions } from 'Types'
import { OptionChooser } from './OptionChooser'
import { DropdownSelections } from '../FiltersUtil'
import { CloudProviderChooser } from './CloudProviderChooser'
import { AccountChooser } from './AccountChooser'
import { ServiceChooser } from './ServiceChooser'

export default function createOptionChooser(
  filterType: DropdownFilterOptions,
  selections: DropdownOption[],
  oldSelections: DropdownSelections,
  filterOptions: FilterOptions,
): OptionChooser {
  switch (filterType) {
    case DropdownFilterOptions.CLOUD_PROVIDERS:
      return new CloudProviderChooser(selections, oldSelections, filterOptions)
    case DropdownFilterOptions.ACCOUNTS:
      return new AccountChooser(selections, oldSelections, filterOptions)
    case DropdownFilterOptions.SERVICES:
      return new ServiceChooser(selections, oldSelections, filterOptions)
  }
}
