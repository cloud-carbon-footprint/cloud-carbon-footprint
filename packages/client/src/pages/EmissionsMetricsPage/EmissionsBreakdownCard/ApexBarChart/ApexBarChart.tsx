/*
 * Â© 2021 Thoughtworks, Inc.
 */

import React, { FunctionComponent, useState } from 'react'
import Chart from 'react-apexcharts'
import { useTheme } from '@material-ui/core/styles'
import { PageEntry, Page, barChartCustomColors } from '../../../../Types'
import useStyles from './apexBarChartStyles'
import { createCustomBarColors, mapToRange } from './helpers'
import Pagination from '../Pagination'
import CarbonIntensityRange from '../CarbonIntensityRange/'
import { EmissionRatioResult } from '@cloud-carbon-footprint/common'

type ApexBarChartProps = {
  data: { string: [string, number] }
  dataType?: string
  emissionsData: EmissionRatioResult[]
}

const ApexBarChart: FunctionComponent<ApexBarChartProps> = ({
  data,
  dataType,
  emissionsData,
}) => {
  const [pageData, setPageData] = useState<Page<PageEntry>>({
    data: [],
    page: 0,
  })

  const theme = useTheme()
  const classes = useStyles()
  const mainTheme = theme.palette.primary.main
  const darkTheme = theme.palette.primary.dark

  let customBarColors = [mainTheme]
  if (dataType === 'region' && !!emissionsData.length) {
    customBarColors = createCustomBarColors(pageData, emissionsData, mainTheme)
  }

  const dataEntries: { x: string[]; y: number }[] = Object.entries(data)
    .filter((item: [string, [string, number]]) => item[1][1] > 0)
    .map((item) => {
      return {
        x: [item[0], `(${item[1][0]})`],
        y: item[1][1],
      }
    })
    .sort((higherC02, lowerCO2) => lowerCO2.y - higherC02.y)

  const smallestCO2E = dataEntries?.[dataEntries?.length - 1]?.y
  const largestCO2E = dataEntries?.[0]?.y
  const totalCO2EByDataType = dataEntries.reduce((acc, currentValue) => {
    return acc + currentValue.y
  }, 0)

  const pageSize = 10
  const minThreshold = 1
  const maxThreshold = 100
  const mappedDataEntries: PageEntry[] = dataEntries.map((entry) => {
    const yEntry = mapToRange(
      entry.y,
      smallestCO2E,
      largestCO2E,
      minThreshold,
      maxThreshold,
    )
    return { x: entry.x, y: isNaN(yEntry) ? maxThreshold : yEntry }
  })

  const toolbarOffset = {
    x: -120,
    y: dataType === 'region' ? -122 : -55,
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const options: any = {
    series: [
      {
        name: 'Total CO2e',
        data: pageData.data,
      },
    ],
    colors: customBarColors,
    chart: {
      type: 'bar',
      toolbar: {
        offsetX: toolbarOffset.x,
        offsetY: toolbarOffset.y,
        tools: {
          download: `
            <div class="apexcharts-menu-icon" title="Menu">
                <svg class="MuiSvgIcon-root" focusable="false" viewBox="0 0 24 24" aria-hidden="true">
                <path d="M19 9h-4V3H9v6H5l7 7 7-7zM5 18v2h14v-2H5z"></path>
                </svg>
            </div>   
           `,
        },
      },
    },
    grid: {
      show: false,
      yaxis: {
        lines: {
          show: false,
        },
      },
      xaxis: {
        lines: {
          show: false,
        },
      },
      padding: {
        left: dataType === 'region' ? -15 : -5,
      },
    },
    plotOptions: {
      bar: {
        horizontal: true,
        barHeight: `${7 * pageData.data.length}%`,
        distributed: true,
      },
    },
    legend: {
      show: false,
    },
    dataLabels: {
      enabled: true,
      textAnchor: 'start',
      formatter: function (_: number, opts: { dataPointIndex: number }) {
        const currentCO2E =
          dataEntries[pageData.page * pageSize + opts.dataPointIndex]?.y

        const formattedPercentage = (currentCO2E / totalCO2EByDataType) * 100
        return formattedPercentage < 0.01
          ? '< 0.01 %'
          : `${formattedPercentage.toFixed(2)} %`
      },
      offsetX: 16,
      background: {
        enabled: true,
        foreColor: mainTheme,
        borderColor: darkTheme,
        padding: 6,
        borderRadius: 1,
        borderWidth: 1,
        opacity: 0.9,
      },
    },
    xaxis: {
      type: 'category',
      labels: {
        style: {
          fontSize: 0,
        },
      },
      axisBorder: {
        show: false,
      },
      max: maxThreshold,
    },
    yaxis: {
      labels: {
        minWidth: 150,
        maxWidth: 150,
        offsetY: 5,
        style: {
          fontSize: '12px',
        },
      },
    },
    tooltip: {
      fillSeriesColor: false,
      y: {
        formatter: function (value: number, opts: { dataPointIndex: number }) {
          return `${dataEntries[
            pageData.page * pageSize + opts.dataPointIndex
          ].y.toFixed(3)} metric tons`
        },
      },
    },
    height: '500px',
  }

  const handlePage = (page: Page<PageEntry>) => {
    setPageData(page)
  }

  return (
    <div className={classes.barChartContainer}>
      {dataType === 'region' && (
        <CarbonIntensityRange
          startLabel="Low carbon intensity"
          endLabel="High carbon intensity"
          colorRange={barChartCustomColors}
        />
      )}
      <Chart
        options={options}
        series={options.series}
        type="bar"
        height={options.height}
      />
      <Pagination
        data={mappedDataEntries}
        pageSize={pageSize}
        handlePage={handlePage}
      />
    </div>
  )
}

export default ApexBarChart
