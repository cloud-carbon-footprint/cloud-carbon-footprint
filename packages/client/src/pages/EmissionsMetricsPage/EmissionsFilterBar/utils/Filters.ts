/*
 * Â© 2021 Thoughtworks, Inc.
 */

import moment from 'moment'
import { EstimationResult } from '@cloud-carbon-footprint/common'
import * as FiltersUtil from './FiltersUtil'
import {
  ALL_ACCOUNTS_DROPDOWN_OPTION,
  ALL_SERVICES_DROPDOWN_OPTION,
  CLOUD_PROVIDER_OPTIONS,
} from './DropdownConstants'
import {
  FilterResultResponse,
  UnknownTypes,
  DropdownOption,
  FilterOptions,
  DropdownFilterOptions,
} from 'Types'

type MaybeDateRange = DateRange | null
type MaybeMoment = moment.Moment | null

interface FiltersConfig {
  timeframe: number
  dateRange: MaybeDateRange
  [DropdownFilterOptions.SERVICES]: DropdownOption[]
  [DropdownFilterOptions.CLOUD_PROVIDERS]: DropdownOption[]
  [DropdownFilterOptions.ACCOUNTS]: DropdownOption[]
}

const defaultFiltersConfig = {
  timeframe: 36,
  dateRange: null,
  [DropdownFilterOptions.SERVICES]: [ALL_SERVICES_DROPDOWN_OPTION],
  [DropdownFilterOptions.CLOUD_PROVIDERS]: CLOUD_PROVIDER_OPTIONS,
  [DropdownFilterOptions.ACCOUNTS]: [ALL_ACCOUNTS_DROPDOWN_OPTION],
}

export const filtersConfigGenerator = ({
  accounts,
  services,
}: FilterResultResponse): FiltersConfig => {
  const configValues = {
    accounts: [ALL_ACCOUNTS_DROPDOWN_OPTION, ...accounts],
    services: [ALL_SERVICES_DROPDOWN_OPTION, ...services],
  }

  return Object.assign(defaultFiltersConfig, configValues)
}

export class Filters {
  readonly timeframe: number
  readonly dateRange: MaybeDateRange
  readonly services: DropdownOption[]
  readonly cloudProviders: DropdownOption[]
  readonly accounts: DropdownOption[]

  constructor(config: FiltersConfig = defaultFiltersConfig) {
    this.timeframe = config.timeframe
    this.dateRange = config.dateRange
    this.services = config.services
    this.cloudProviders = config.cloudProviders
    this.accounts = config.accounts
  }

  withTimeFrame(timeframe: number): Filters {
    return this.timeframe === timeframe
      ? this
      : new Filters({ ...this, dateRange: null, timeframe })
  }

  withDropdownOption(
    dropdownOption: DropdownOption[],
    filterOptions: FilterOptions,
    dropdownFilter: DropdownFilterOptions,
  ): Filters {
    const oldSelections = {
      services: this.services,
      accounts: this.accounts,
      cloudProviders: this.cloudProviders,
    }
    return new Filters({
      ...this,
      ...FiltersUtil.handleDropdownSelections(
        dropdownFilter,
        dropdownOption,
        oldSelections,
        filterOptions,
      ),
    })
  }

  // withServices(
  //   services: DropdownOption[],
  //   filterOptions: FilterOptions,
  // ): Filters {
  //   const oldSelections = {
  //     services: this.services,
  //     accounts: this.accounts,
  //     cloudProviders: this.cloudProviders,
  //   }
  //   return new Filters({
  //     ...this,
  //     ...FiltersUtil.handleDropdownSelections(
  //       FiltersUtil.DropdownFilter.SERVICES,
  //       services,
  //       oldSelections,
  //       filterOptions,
  //     ),
  //   })
  // }
  //
  // withAccounts(
  //   accounts: DropdownOption[],
  //   filterOptions: FilterOptions,
  // ): Filters {
  //   const oldSelections = {
  //     services: this.services,
  //     accounts: this.accounts,
  //     cloudProviders: this.cloudProviders,
  //   }
  //   return new Filters({
  //     ...this,
  //     ...FiltersUtil.handleDropdownSelections(
  //       FiltersUtil.DropdownFilter.ACCOUNTS,
  //       accounts,
  //       oldSelections,
  //       filterOptions,
  //     ),
  //   })
  // }
  //
  // withCloudProviders(
  //   cloudProviders: DropdownOption[],
  //   filterOptions: FilterOptions,
  // ): Filters {
  //   const oldSelections = {
  //     services: this.services,
  //     accounts: this.accounts,
  //     cloudProviders: this.cloudProviders,
  //   }
  //   return new Filters({
  //     ...this,
  //     ...FiltersUtil.handleDropdownSelections(
  //       FiltersUtil.DropdownFilter.CLOUD_PROVIDERS,
  //       cloudProviders,
  //       oldSelections,
  //       filterOptions,
  //     ),
  //   })
  // }

  withDateRange(dateRange: DateRange): Filters {
    if (dateRange.isComplete()) {
      return new Filters({
        ...this,
        timeframe: -1,
        dateRange,
      })
    } else {
      return new Filters({
        ...this,
        dateRange,
      })
    }
  }

  serviceLabel(serviceOptions: DropdownOption[]): string {
    return FiltersUtil.numSelectedLabel(
      this.services.length,
      serviceOptions.length,
    )
  }

  cloudProviderLabel(): string {
    return FiltersUtil.numSelectedLabel(
      this.cloudProviders.length,
      CLOUD_PROVIDER_OPTIONS.length,
      'Cloud Providers',
    )
  }

  accountLabel(accountOptions: DropdownOption[]): string {
    return FiltersUtil.numSelectedLabel(
      this.accounts.length,
      accountOptions.length,
      'Accounts',
    )
  }

  filter(rawResults: EstimationResult[]): EstimationResult[] {
    const resultsFilteredByTime = this.getResultsFilteredByTime(rawResults)
    const resultsFilteredByService = this.getResultsFilteredByService(
      resultsFilteredByTime,
    )
    return this.getResultsFilteredByAccount(resultsFilteredByService)
  }

  private getResultsFilteredByAccount(
    resultsFilteredByService: EstimationResult[],
  ): EstimationResult[] {
    const allAccountsSelected = this.accounts.includes(
      ALL_ACCOUNTS_DROPDOWN_OPTION,
    )
    return resultsFilteredByService
      .map((estimationResult) => {
        const filteredServiceEstimates =
          estimationResult.serviceEstimates.filter((serviceEstimate) => {
            return (
              this.accounts.some(
                (account) =>
                  (account.name.includes(UnknownTypes.UNKNOWN_ACCOUNT) &&
                    serviceEstimate.accountName === null) ||
                  account.name === serviceEstimate.accountName,
              ) || allAccountsSelected
            )
          })
        return {
          timestamp: estimationResult.timestamp,
          serviceEstimates: filteredServiceEstimates,
        }
      })
      .filter(
        (estimationResult) => !!estimationResult?.serviceEstimates?.length,
      )
  }

  private getResultsFilteredByService(
    resultsFilteredByTime: EstimationResult[],
  ): EstimationResult[] {
    const allServicesSelected = this.services.includes(
      ALL_SERVICES_DROPDOWN_OPTION,
    )
    return resultsFilteredByTime.map((estimationResult) => {
      const filteredServiceEstimates = estimationResult.serviceEstimates.filter(
        (serviceEstimate) => {
          return (
            this.services.some(
              (service) =>
                (service.key.includes(UnknownTypes.UNKNOWN_SERVICE) &&
                  serviceEstimate.serviceName === null) ||
                service.key === serviceEstimate.serviceName,
            ) || allServicesSelected
          )
        },
      )
      return {
        timestamp: estimationResult.timestamp,
        serviceEstimates: filteredServiceEstimates,
      }
    })
  }

  private getResultsFilteredByTime(
    rawResults: EstimationResult[],
  ): EstimationResult[] {
    const today = moment.utc()
    let start: moment.Moment
    let end: moment.Moment
    if (this.timeframe < 0 && this.dateRange?.startDate) {
      start = this.dateRange.startDate
      end = this.dateRange.endDate || today
    } else {
      end = today
      start = end.clone().subtract(this.timeframe, 'M')
    }

    return rawResults.filter((estimationResult: EstimationResult) =>
      moment.utc(estimationResult.timestamp).isBetween(start, end, 'day', '[]'),
    )
  }
}

export class DateRange {
  readonly startDate: MaybeMoment
  readonly endDate: MaybeMoment

  constructor(startDate: MaybeMoment, endDate: MaybeMoment) {
    this.startDate = startDate
    this.endDate = endDate
  }

  isComplete(): boolean {
    return this.startDate !== null && this.endDate !== null
  }
}
