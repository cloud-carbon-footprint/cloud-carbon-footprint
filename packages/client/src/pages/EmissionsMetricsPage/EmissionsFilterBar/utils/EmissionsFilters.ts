/*
 * Â© 2021 Thoughtworks, Inc.
 */

import moment from 'moment'
import { Filters } from '../../../../common/FilterBar/utils/Filters'
import {
  DropdownFilterOptions,
  DropdownOption,
  FilterOptions,
  FilterResultResponse,
  FiltersConfig,
  unknownOptionTypes,
} from '../../../../Types'
import {
  ALL_ACCOUNTS_DROPDOWN_OPTION,
  ALL_DROPDOWN_FILTER_OPTIONS,
  ALL_SERVICES_DROPDOWN_OPTION,
  CLOUD_PROVIDER_OPTIONS,
} from '../../../../common/FilterBar/utils/DropdownConstants'
import { EstimationResult, ServiceData } from '@cloud-carbon-footprint/common'
import { AccountChooser } from './options/AccountChooser'
import { DropdownSelections } from '../../../../common/FilterBar/utils/FiltersUtil'
import { OptionChooser } from '../../../../common/FilterBar/utils/OptionChooser'
import { CloudProviderChooser } from './options/CloudProviderChooser'
import { ServiceChooser } from './options/ServiceChooser'

const defaultConfig: FiltersConfig = {
  timeframe: 36,
  dateRange: null,
  options: {
    [DropdownFilterOptions.SERVICES]: [ALL_SERVICES_DROPDOWN_OPTION],
    [DropdownFilterOptions.CLOUD_PROVIDERS]: CLOUD_PROVIDER_OPTIONS,
    [DropdownFilterOptions.ACCOUNTS]: [ALL_ACCOUNTS_DROPDOWN_OPTION],
  },
}

export class EmissionsFilters extends Filters {
  constructor(config = defaultConfig) {
    super(config)
  }

  static generateConfig(newOptions: FilterResultResponse): FiltersConfig {
    return super.generateConfig(newOptions, defaultConfig)
  }

  create(config: FiltersConfig): Filters {
    return new EmissionsFilters(config)
  }

  createOptionChooser(
    filterType: DropdownFilterOptions,
    selections: DropdownOption[],
    oldSelections: DropdownSelections,
    filterOptions: FilterOptions,
  ): OptionChooser {
    switch (filterType) {
      case DropdownFilterOptions.CLOUD_PROVIDERS:
        return new CloudProviderChooser(
          selections,
          oldSelections,
          filterOptions,
        )
      case DropdownFilterOptions.ACCOUNTS:
        return new AccountChooser(selections, oldSelections, filterOptions)
      case DropdownFilterOptions.SERVICES:
        return new ServiceChooser(selections, oldSelections, filterOptions)
    }
  }

  filter(rawResults: EstimationResult[]): EstimationResult[] {
    const resultsFilteredByTime = this.getResultsFilteredByTime(rawResults)
    const resultsFilteredByService = this.getResultsFilteredBy(
      DropdownFilterOptions.SERVICES,
      resultsFilteredByTime,
    )
    return this.getResultsFilteredBy(
      DropdownFilterOptions.ACCOUNTS,
      resultsFilteredByService,
    ).filter((estimationResult) => !!estimationResult?.serviceEstimates?.length)
  }

  private getResultsFilteredByTime(
    previousResults: EstimationResult[],
  ): EstimationResult[] {
    const today = moment.utc()
    let start: moment.Moment
    let end: moment.Moment
    if (this.timeframe < 0 && this.dateRange?.startDate) {
      start = this.dateRange.startDate
      end = this.dateRange.endDate || today
    } else {
      end = today
      start = end.clone().subtract(this.timeframe, 'M')
    }

    return previousResults.filter((estimationResult: EstimationResult) =>
      moment.utc(estimationResult.timestamp).isBetween(start, end, 'day', '[]'),
    )
  }

  private getResultsFilteredBy(
    type: string,
    previousResults: EstimationResult[],
  ): EstimationResult[] {
    const hasAllOptionsSelected = this.options[type].includes(
      ALL_DROPDOWN_FILTER_OPTIONS[type],
    )
    return previousResults.map((estimationResult) => {
      const filteredServiceEstimates = estimationResult.serviceEstimates.filter(
        (serviceEstimate) =>
          this.options[type].some((dropdownOption) =>
            this.isUnknownOrSameName(dropdownOption, type, serviceEstimate),
          ) || hasAllOptionsSelected,
      )
      return {
        timestamp: estimationResult.timestamp,
        serviceEstimates: filteredServiceEstimates,
        periodStartDate: estimationResult.periodStartDate,
        periodEndDate: estimationResult.periodEndDate,
        groupBy: estimationResult.groupBy,
      }
    })
  }

  private isUnknownOrSameName(
    dropdownOption: DropdownOption,
    type: string,
    serviceEstimate: ServiceData,
  ) {
    const name =
      type === DropdownFilterOptions.ACCOUNTS ? 'accountName' : 'serviceName'
    return (
      (dropdownOption.name.includes(unknownOptionTypes[type]) &&
        serviceEstimate[name] === null) ||
      dropdownOption.name === serviceEstimate[name]
    )
  }
}
