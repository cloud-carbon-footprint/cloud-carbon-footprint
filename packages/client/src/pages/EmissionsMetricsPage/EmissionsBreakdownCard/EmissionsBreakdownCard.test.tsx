/*
 * Â© 2021 Thoughtworks, Inc.
 */

import React from 'react'
import {
  act,
  create,
  ReactTestInstance,
  ReactTestRenderer,
} from 'react-test-renderer'
import { EmissionRatioResult } from '@cloud-carbon-footprint/common'
import { ServiceResult } from '../../../Types'
import SelectDropdown from '../../../common/SelectDropdown'
import NoDataMessage from '../../../common/NoDataMessage'
import { useRemoteEmissionService } from '../../../utils/hooks'
import {
  fakeEmissionFactors,
  mockDataWithHigherPrecision,
} from '../../../utils/data'
import EmissionsBreakdownCard from './EmissionsBreakdownCard'
import ApexBarChart from './ApexBarChart'
import { waitFor } from '@testing-library/react'

jest.mock('apexcharts')
jest.mock('../../../utils/themes')
jest.mock('../../../utils/hooks/EmissionFactorServiceHook')

const mockedUseEmissionFactorService =
  useRemoteEmissionService as jest.MockedFunction<
    typeof useRemoteEmissionService
  >

describe('EmissionsBreakdownCard', () => {
  let testRenderer: ReactTestRenderer, testInstance: ReactTestInstance
  const styleClass = 'test-style-class'

  beforeEach(() => {
    const mockReturnValue: ServiceResult<EmissionRatioResult> = {
      loading: false,
      data: fakeEmissionFactors,
    }
    mockedUseEmissionFactorService.mockReturnValue(mockReturnValue)
    testRenderer = create(
      <EmissionsBreakdownCard
        containerClass={styleClass}
        data={mockDataWithHigherPrecision}
        baseUrl="/api"
      />,
    )
    testInstance = testRenderer.root
  })

  afterEach(() => {
    testRenderer.unmount()
    mockedUseEmissionFactorService.mockClear()
  })

  it('fetches emissions factors', async () => {
    await waitFor(() => {
      expect(mockedUseEmissionFactorService).toHaveBeenCalledWith({
        baseUrl: '/api',
        onApiError: undefined,
      })
    })
  })

  it('renders a select dropdown for the bar chart', () => {
    const selectDropdownInstances = testInstance.findAllByType(SelectDropdown)

    expect(selectDropdownInstances).toHaveLength(1)
    expect(testRenderer.toJSON()).toMatchSnapshot()
  })

  it('renders the apex bar chart to display data', () => {
    const isApexBarChartRendered = testInstance.findAllByType(ApexBarChart)

    expect(isApexBarChartRendered).toHaveLength(1)
  })

  it('renders emission breakdowns by region on the bar chart by default', () => {
    const isApexBarChartRendered = testInstance.findByType(ApexBarChart)

    expect(isApexBarChartRendered.props.dataType).toBe('region')
  })

  it('updates the chart type when a different one is selected from the dropdown', () => {
    act(() =>
      testInstance
        .findByType(SelectDropdown)
        .props.handleChange({ target: { value: 'Service' } }),
    )

    expect(testInstance.findByType(ApexBarChart).props.dataType).toBe('Service')
  })

  it('should show a no data message when there is no data to display', () => {
    testRenderer.unmount()
    const fixture = create(<EmissionsBreakdownCard data={[]} />)
    const noDataComponent = fixture.root.findByType(NoDataMessage)

    expect(noDataComponent).toBeDefined()
    expect(noDataComponent.props.title).toBe('Emissions Breakdown')
  })
})
