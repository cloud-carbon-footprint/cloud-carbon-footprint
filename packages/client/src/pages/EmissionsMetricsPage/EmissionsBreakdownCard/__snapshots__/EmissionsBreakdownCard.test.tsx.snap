// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmissionsBreakdownCard renders a select dropdown for the bar chart 1`] = `
<div
  className="MuiGrid-root makeStyles-cardHalf-17 MuiGrid-item"
>
  <div
    className="MuiPaper-root MuiCard-root makeStyles-card-16 makeStyles-minHeight-18 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      className="MuiBox-root MuiBox-root-22 makeStyles-contentContainer-20"
    >
      <div
        className="MuiCardContent-root makeStyles-cardContent-21"
      >
        <div
          className="makeStyles-topContainer-14"
          id="emissionsBreakdownContainer"
        >
          <p
            className="makeStyles-title-15"
          >
            Emissions Breakdown
          </p>
          <div
            className="MuiFormControl-root"
          >
            <div
              className="MuiInputBase-root MuiInputBase-formControl"
              data-testid="select"
              onClick={[Function]}
            >
              <div
                aria-haspopup="listbox"
                aria-labelledby="breakdown-selector"
                className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input WithStyles(ForwardRef(InputBase))-input-23"
                id="breakdown-selector"
                onBlur={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onMouseDown={[Function]}
                role="button"
                tabIndex={0}
              >
                Region
              </div>
              <input
                aria-hidden={true}
                className="MuiSelect-nativeInput"
                onAnimationStart={[Function]}
                onChange={[Function]}
                required={false}
                tabIndex={-1}
                value="region"
              />
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          className="makeStyles-barChartContainer-24"
        >
          <div
            style={
              Object {
                "alignSelf": "center",
                "display": "flex",
                "flexDirection": "column",
                "padding": "10px 45px",
                "width": "100%",
              }
            }
          >
            <div
              style={
                Object {
                  "display": "flex",
                  "fontWeight": 500,
                  "justifyContent": "space-between",
                  "paddingBottom": "12px",
                }
              }
            >
              <div>
                Low carbon intensity
              </div>
              <div>
                High carbon intensity
              </div>
            </div>
            <div
              style={
                Object {
                  "alignItems": "center",
                  "display": "flex",
                  "justifyContent": "space-between",
                  "width": "100%",
                }
              }
            >
              <div
                className="t-dot"
                data-testid="timeline-dot"
                style={
                  Object {
                    "backgroundColor": "#73B500",
                    "borderRadius": "8px",
                    "height": "15px",
                    "width": "15px",
                  }
                }
              />
              <div
                className="t-bar"
                style={
                  Object {
                    "backgroundColor": "#aaaaaa",
                    "flexGrow": 2,
                    "height": "1px",
                  }
                }
              />
              <div
                className="t-dot"
                data-testid="timeline-dot"
                style={
                  Object {
                    "backgroundColor": "#00791E",
                    "borderRadius": "8px",
                    "height": "15px",
                    "width": "15px",
                  }
                }
              />
              <div
                className="t-bar"
                style={
                  Object {
                    "backgroundColor": "#aaaaaa",
                    "flexGrow": 2,
                    "height": "1px",
                  }
                }
              />
              <div
                className="t-dot"
                data-testid="timeline-dot"
                style={
                  Object {
                    "backgroundColor": "#D99200",
                    "borderRadius": "8px",
                    "height": "15px",
                    "width": "15px",
                  }
                }
              />
              <div
                className="t-bar"
                style={
                  Object {
                    "backgroundColor": "#aaaaaa",
                    "flexGrow": 2,
                    "height": "1px",
                  }
                }
              />
              <div
                className="t-dot"
                data-testid="timeline-dot"
                style={
                  Object {
                    "backgroundColor": "#DF5200",
                    "borderRadius": "8px",
                    "height": "15px",
                    "width": "15px",
                  }
                }
              />
              <div
                className="t-bar"
                style={
                  Object {
                    "backgroundColor": "#aaaaaa",
                    "flexGrow": 2,
                    "height": "1px",
                  }
                }
              />
              <div
                className="t-dot"
                data-testid="timeline-dot"
                style={
                  Object {
                    "backgroundColor": "#790000",
                    "borderRadius": "8px",
                    "height": "15px",
                    "width": "15px",
                  }
                }
              />
            </div>
          </div>
          <div />
          <div
            className="makeStyles-paginationContainer-25"
          >
            <span
              className="makeStyles-paginationLabel-26"
            >
              1 - 1
               of 
              1
            </span>
            <button
              aria-label="first page"
              className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary Mui-disabled Mui-disabled"
              disabled={true}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={-1}
              type="button"
            >
              <span
                className="MuiIconButton-label"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M18.41 16.59L13.82 12l4.59-4.59L17 6l-6 6 6 6zM6 6h2v12H6z"
                  />
                </svg>
              </span>
            </button>
            <button
              aria-label="chevron-left"
              className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary Mui-disabled Mui-disabled"
              disabled={true}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={-1}
              type="button"
            >
              <span
                className="MuiIconButton-label"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                  />
                </svg>
              </span>
            </button>
            <button
              aria-label="chevron-right"
              className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary Mui-disabled Mui-disabled"
              disabled={true}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={-1}
              type="button"
            >
              <span
                className="MuiIconButton-label"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                  />
                </svg>
              </span>
            </button>
            <button
              aria-label="last page"
              className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary Mui-disabled Mui-disabled"
              disabled={true}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={-1}
              type="button"
            >
              <span
                className="MuiIconButton-label"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M5.59 7.41L10.18 12l-4.59 4.59L7 18l6-6-6-6zM16 6h2v12h-2z"
                  />
                </svg>
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
